raft-tech/arcadedb
# 2289 enforce data classification tagging

## What does this PR do?
Forces ordinary users to provide classification data with new writes, except for service accounts. Ordinary users won't have visibility of documents without proper classification markings, but users with the data steward role will be able to view and manage those documents without proper classification markings.

## Motivation
SOCOM data handling requirements

## Related issues
A list of issues either fixed, containing architectural discussions, otherwise relevant
for this Pull Request.

## Additional Notes
Anything else we should know when reviewing?

## Setup
1. Using the corresponding data fabric PR, spin up your local environment
https://github.com/raft-tech/data-fabric/pull/2334

2. From the root project, run

        docker buildx build --platform linux/arm64 --load -t raft-tech/arcadedev ./

3. Load into kind

        kind load docker-image raft-tech/arcadedev  --name data-fabric
        
4. Update image in arcade's stateful set
5. Kill the pod to force it to update if needed
6. Create a new keycloak user named "service-account-test". Assign to "Arcade Admin" group.
7. In keycloak, remove the Data Steward group from the admin user.

## Testing

1. Log into arcade http://arcadedb.localhost
2. Create a new database through the gui (click the database icon in the upper left, and look at the menu in the upper right)

3. Confirm you can switch to the new database in the database dropdown in the upper middle left of the screen

4. Run the following setup scripts individually

        CREATE VERTEX TYPE Customer

        CREATE DOCUMENT TYPE Classification IF NOT EXISTS

5. Run the following command, confirm it fails with a classification error

        INSERT INTO People CONTENT { firstName: 'Enzo', lastName: 'Ferrari' }

8. Run the following command, confirm it succeeds

        INSERT INTO People CONTENT { firstName: 'Enzo', lastName: 'Ferrari', classification: { @type: 'Classification', general: 'U' } }

9. Run the following command, confirm it fails with an invalid classification type error

        INSERT INTO People CONTENT { firstName: 'Enzo', lastName: 'Ferrari', classification: { @type: 'Classification', general: 'A' } }

10. Open a new session of arcade, and sign in as the "service-account-test' user.

11. Run the following command, confirm it succeeds

        INSERT INTO People CONTENT { firstName: 'Enzo2', lastName: 'Ferrari' }

12. Confirm the `classificationMarked` field is set to false. It may help to change the data display from graph to table or json.

13. Run the following command, confirm you can see both documents we created earlier, with 1 each of classification true or false.

        select from `People` limit 30

14. Open a new session of arcade, log in as admin user

15. Rerun the select command above. Confirm you only see a single document, the one with classification info and marked as true.

16. Sign out of arcade

17. In keycloak, add the admin user to the Data Steward Admin group.

18. Re-run the select command above, confirm you see both docs.
